### PR TITLE
Re-Implementing Sudoku Board Component

### PR DESCRIPTION
**Pull request by Thomas-Gallant**
> [!IMPORTANT] 
> Temporarily removing Hint logic, Drill logic, and Demo logic for re-implementation
> Test Cases will fail, will be fixed in future PRs

## Changelog

- moving metro-runtime to top of App.tsx so that fast-refresh works
- Moving all images to `.assets` folder
- Simplifying folder structure
- Changed Sudoku Board to use new SudokuBoard object 
- Adding project recommended extensions
- Removed duplicate FormatTime functions
- Can unselect cell
- move history preserved through game refreshes, stored in game object
- can use "0" key to erase value/notes
- board renders from rows left to right, top to bottom, instead of rendering columns top to bottom, left to right. 
- erase button disabled if cell not selected
- disable erase button if selected cell is empty, is given, or is correct
- disable numpad buttons if value is correct
- disable numpad buttons if no value is selected
- Incorrect cell highlighted even when no cells are selected.
- Fixed bug where timer stops on state update
- Removed EndGame modal, replacing with EndGame screen
- All ts-ignores have been removed.
- Adding more types
- startGame, saveGame, finishGame, and resumeGame functions updated to use new object structure
- Switches to retrieve final statistics from memory state instead of localstorage
- finishGame score calculations moved outside of function, statistics provided to finishGame function instead of calculated inside it
- Adding interfaces for all Sudoku Board subcomponents
- removing all stylesheets
- endGame only evaluated on value entry, and not on every render of Cell component
- Object structure temporarily defined in LocalDatabase.ts
- Adding documentation for Board Component
- removing immutable and prop-types dependencies

## Checklist for completing pull request:
- [ ] Test functionality on Mobile, Web, and Desktop
- [ ] Verify that any tests for new functionality are created and functional.
- [ ] If needed, update the Readme

### CODE REVIEW

File: docs/Components/SudokuBoard/BoardVariants.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Board Variants
+
+## Classic Variant
+
+Sudoku game with 9x9 grid.
+
+## Drill Variant
+
+Sudoku game with 9x9 grid. Garenteed to have a certain strategy avaliable as a next move. The user must follow the correct steps to solve the strategy and complete the drill.
```
Comment (Gregory711) [inline docs/Components/SudokuBoard/BoardVariants.md:9]: nit: guaranteed is misspelled

Comment (Thomas-Gallant) [reply docs/Components/SudokuBoard/BoardVariants.md:9]: Thanks! I updated

File: docs/Components/SudokuBoard/BoardLogic.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# Classic Variant
+
+## Insertion Logic
+
+```mermaid
```
Comment (Gregory711) [inline docs/Components/SudokuBoard/BoardLogic.md:5]: Awesome use of a mermaid chart for documentation! I do have a nit: would be nice if you used [this format](https://mermaid.js.org/syntax/flowchart.html#graph-declarations-with-spaces-between-vertices-and-link-and-without-semicolon) of having a diamond shape connected to the User Selects Cell which says something about the state of that pencil button (not sure if you have an official term for it in docs already?) and then the lines to the next two parts have text on them declaring the state of the pencil that causes each path to be taken.

File: app/Components/Sudoku Board/Functions/BoardFunctions.ts [hunk 3]
```diff
@@ -186,159 +61,48 @@ export const formatTime = (inputSeconds: number) => {
...
+): number {
+  const numericalDifficulty: GameDifficultyScore =
+    difficulty === "easy"
+      ? 10
+      : difficulty === "medium"
+      ? 20
+      : difficulty === "hard"
+      ? 30
+      : 10;
+  // calculate score
```
Comment (Gregory711) [inline app/Components/Sudoku Board/Functions/BoardFunctions.ts:95]: nit: the score calculation uses a lot of ternary operators and doesn't have any comments making it hard to read. Also I can't tell if numericalDifficulty / 1000 will be equal to 0 or 0.01 for 10 cause of float vs int. Would be nice to have some comments explaining how this all works and maybe an example score calculation

File: docs/Components/SudokuBoard/BoardLogic.md (hunk 2 above)
Comment (Thomas-Gallant) [reply docs/Components/SudokuBoard/BoardLogic.md:5]: Do you mean something like this?

```mermaid
flowchart
    PencilButtonState{{Note Button State}}
    InsertStart[User Selects Cell]
    InsertValue[User Presses HotKey/Button <br> to insert value into Cell]
    InsertNote[User Presses HotKey/Button <br>to insert note into Cell]
    PushHistory[Board pushes existing cell state <br> to ActionHistory array]
    UpdateCell[Cell is updated with new value]

    PencilButtonState --> InsertStart
    InsertStart --> |If Note button is disabled|InsertValue
    InsertStart --> |If Note button is enabled|InsertNote
    InsertValue --> |If Insert is Valid|PushHistory
    InsertNote --> |If Insert is Valid|PushHistory
    PushHistory --> UpdateCell
```

File: app/Components/Sudoku Board/sudoku.ts [hunk 4]
```diff
@@ -1,69 +1,58 @@
...
+    Math.floor(currentCellCoordinate.r / 3);
+  const squareB =
+    Math.floor(selectedCellCoordinate.c / 3) * 3 +
+    Math.floor(selectedCellCoordinate.r / 3);
   return squareA === squareB;
 }
 
-export function highlightRow(a, b) {
-  return a.y === b.y;
+export function highlightRow(
```
Comment (Gregory711) [inline app/Components/Sudoku Board/sudoku.ts:44]: I don't understand this function

Comment (Gregory711) [reply app/Components/Sudoku Board/sudoku.ts:44]: Maybe if you could add a comment at the top describing the inputs and output?

File: app/Components/Sudoku Board/Functions/BoardFunctions.ts (hunk 3 above)
Comment (Thomas-Gallant) [reply app/Components/Sudoku Board/Functions/BoardFunctions.ts:95]: Good call, I'll work on docs/comments for that and clean up that code a bit

File: app/Components/Sudoku Board/sudoku.ts [hunk 5]
```diff
@@ -1,69 +1,58 @@
...
-export function highlightBox(a, b) {
-  const squareA = Math.floor(a.x / 3) * 3 + Math.floor(a.y / 3);
-  const squareB = Math.floor(b.x / 3) * 3 + Math.floor(b.y / 3);
+export function highlightBox(
+  currentCellCoordinate: CellLocation,
+  selectedCellCoordinate: CellLocation
+) {
+  const squareA =
+    Math.floor(currentCellCoordinate.c / 3) * 3 +
+    Math.floor(currentCellCoordinate.r / 3);
```
Comment (Gregory711) [inline app/Components/Sudoku Board/sudoku.ts:37]: You use this logic repeatedly in this file. Could you create a function to calculate the box index?

File: docs/Components/SudokuBoard/BoardLogic.md (hunk 2 above)
Comment (Gregory711) [reply docs/Components/SudokuBoard/BoardLogic.md:5]: Almost, basically just flip the Note Button State and the User Select Cell (seems to be a standard style to use diamond shape to do decision/if logic)
image

File: app/Components/Sudoku Board/sudoku.ts [hunk 6]
```diff
@@ -1,69 +1,58 @@
...
+  const squareA =
+    Math.floor(selectedCellCoordinate.r / 3) * 3 +
+    Math.floor(selectedCellCoordinate.c / 3);
+  const squareB =
+    Math.floor(currentCellCoordinate.r / 3) * 3 +
+    Math.floor(currentCellCoordinate.c / 3);
+  return (
+    selectedCellCoordinate.r === currentCellCoordinate.r ||
+    selectedCellCoordinate.c === currentCellCoordinate.c ||
+    squareA === squareB
```
Comment (Gregory711) [inline app/Components/Sudoku Board/sudoku.ts:27]: This function seems to repeat the work done by highlightBox/RowColumn, why not use those?

Comment (Gregory711) [reply app/Components/Sudoku Board/sudoku.ts:27]: Also, like highlightBox/Row/Column would be nice to have a comment above the function to describe what it does, inputs, and outputs

Comment (Thomas-Gallant) [reply app/Components/Sudoku Board/sudoku.ts:27]: It turns out we are not using this function in the code, I'll go ahead and remove it.

File: docs/Components/SudokuBoard/BoardLogic.md (hunk 2 above)
Comment (Thomas-Gallant) [reply docs/Components/SudokuBoard/BoardLogic.md:5]: I made this change, and also updated the Undo logic diagram to have the diamond shape as well for the "User enters undo" decision.

**Review by Gregory711 — APPROVED** [review]

**Review by Gregory711 — APPROVED** [review]

**Pull request merged**